### ISSUE TITLE
Non-standard locations in tree for DetailAST items

### ISSUE THREAD

**Issue by rnveach** [checkstyle/checkstyle#950]
When traversing the DetailAST tree to build a custom checker, I notice some of the same types are found in different areas, either as a child or as a sibling. One example is "return" has ";" as a child while an expression has ";" as a sibling.
It is making traversing the tree and finding items a little bit more tricky than it is needed. It would be much easier if they were to be found in the same location. 

I am using checkstyle 6.5 and extending the "Check" class and getting the DetailAST from the "visitToken" method.
Here is an example I am seeing. Code:

``````
// line 9
	if (a == 1)
		return;
// line 15
	if (a == 2)
		b = 3;
``````

produces the following DetailAST tree:

```````
Type "if" (83) Line 9
	Type "(" (76) Line 9
	Type "EXPR" (28) Line 9
		Type "==" (116) Line 9
			Type "a" (58) Line 9
			Type "1" (137) Line 9
	Type ")" (77) Line 9
	Type "return" (88) Line 9
		Type ";" (45) Line 9 // <-- child of "return"

Type "if" (83) Line 15
	Type "(" (76) Line 15
	Type "EXPR" (28) Line 15
		Type "==" (116) Line 15
			Type "a" (58) Line 15
			Type "2" (137) Line 15
	Type ")" (77) Line 15
	Type "EXPR" (28) Line 16
		Type "=" (80) Line 16
			Type "b" (58) Line 16
			Type "3" (137) Line 16
	Type ";" (45) Line 16 // <-- sibling of "EXPR"
``````

Another example is where {}s are children and ()s for if statements are siblings.

**Comment by romani** [comment]
Thanks a lot for reporting that!

>Another example is where {}s are children and ()s for if statements are siblings.

Please ptovide code example and how tree is looks like, you can do that easily by our gui swing application , and the attach a screenshot. How to run gui is there http://checkstyle.sourceforge.net/cmdline.html#Download_and_Run

>It would be much easier if they were to be found in the same location.

Sure, please do not wait for quick fix, as it require update of grammar and then retesting of all Checks that they are ok with tree structure changes. That is not simple as it could look like.

**Comment by rnveach** [comment]
I realize the level of rewriting this will most likely take and how it will affect people's existing custom checks.

Code:
````````
		if ((a == 2) || (a == 1)) {
			return;
		}
````````
![brace](https://cloud.githubusercontent.com/assets/5427943/7222171/1caccb36-e6d6-11e4-9e4c-8b65fc569636.png)

End brace is a child of start brace while end parenthesis is a sibling of start parenthesis.

**Comment by rnveach** [comment]
Single line comments seems like it falls in this list too.
I have a comment above java code, not in the middle of a statement, yet the parser is putting it in with the METHOD_CALL of the next statement, making it trickier to track down for validation. If there was no java statements below the comment, it would be a first level child of SLIST instead of a deeper child.
I found this out when trying to provide comment overrides for catches like EmptyCatchBlockCheck does. It's method "getCommentFirstLine" will not get the comment in this scenario, even though there are in the same positions in the text code.

code:
``````
try {
	....
}
catch (SQLBuilderException e) {
	// ignore exception, print standard error

	enterpriseTO.setReturnCode(ReturnCode.RELEASE_ERROR);
}
``````

What the parser tree looks like:
``````
Type "catch" (LITERAL_CATCH) Line 2426 Column 4
	Type "(" (LPAREN) Line 2426 Column 10
	Type "PARAMETER_DEF" (PARAMETER_DEF) Line 2426 Column 11
		Type "MODIFIERS" (MODIFIERS) Line 2426 Column 11
		Type "TYPE" (TYPE) Line 2426 Column 11
			Type "SQLBuilderException" (IDENT) Line 2426 Column 11
		Type "e" (IDENT) Line 2426 Column 31
	Type ")" (RPAREN) Line 2426 Column 32
	Type "{" (SLIST) Line 2426 Column 34
		Type "EXPR" (EXPR) Line 2429 Column 31
			Type "(" (METHOD_CALL) Line 2429 Column 31
				Type "." (DOT) Line 2429 Column 17
					Type "//" (SINGLE_LINE_COMMENT) Line 2427 Column 5 // <-- hidden child
						Type " ignore exception, print standard error" (COMMENT_CONTENT) Line 2427 Column 7
					Type "enterpriseTO" (IDENT) Line 2429 Column 5
					Type "setReturnCode" (IDENT) Line 2429 Column 18
				Type "ELIST" (ELIST) Line 2429 Column 42
					Type "EXPR" (EXPR) Line 2429 Column 42
						Type "." (DOT) Line 2429 Column 42
							Type "ReturnCode" (IDENT) Line 2429 Column 32
							Type "RELEASE_ERROR" (IDENT) Line 2429 Column 43
				Type ")" (RPAREN) Line 2429 Column 56
		Type ";" (SEMI) Line 2429 Column 57
		Type "}" (RCURLY) Line 2431 Column 4
``````

The class TokenTypes, which has been a great help, does not seem to note the weird locations comments can come up in the structure. I assume it is because there are too many possibilities.

**Comment by nmancus1** [comment]
If we still want to pursue this issue, it should be broken into issues based on individual tokens , i.e. `SEMI token placement in AST is not consistent`.

Now that we use the visitor, it will be fairly simple to fix these sorts of issues (other than related check updates).

**Comment by rnveach** [comment]
> End brace is a child of start brace while end parenthesis is a sibling of start parenthesis.

@romani Agreed to not touch expressions as changing them will lose order of operation.
